### DISCUSSION TITLE
Storing JSON in string format as dynamic property without deserializing then serializing - C# client

### DISCUSSION THREAD

**Discussion by ben-jacobs** [ravendb/ravendb#13627] [Q&A] [answered]
Hi everyone - It's been a long day, I'm having brain fade and I'm going around in circles. There's been a few different approaches I've used over the history of RavenDB and I guess I've just confused myself. Hopefully someone can help point me in the right direction without much bother.

Per the subject I'd like to create a document using a C# class (ie the normal way) in RavenDB 5.x that contains a property that is essentially a dynamic/JObject when it gets retrieved, but is currently a JSON object in string format (i.e. I don't want it strongly-typed), but I also don't want to store it in "string" format. The outer document is pretty simple and static - i.e. can be a templated string if necessary.

This whole process is super easy if I just deserialize the string and set a property of type dynamic (JObject.Parse) ... job done etc. etc. But that step is really wasteful and adds a heap of time over large numbers of documents.

I _think_ my previous approach was to essentially serialize my c# object, inject the JSON string into the serialized C# object (in fact string templating/interpolation means I don't even need to serialize the c# class each time), then use Blittable(something) to write the string as the document contents. 

All the other approaches I could think of off the top of my head involved de-serializing/serializing wastefully somewhere along the line.

Does anyone have any little tricks they've used to do this sort of thing efficiently in recent versions?

Ben

**Comment by omerlewitz** [comment]
Hi @ben-jacobs 

Document in RavenDB is schemaless so you don't need string property to achieve dynamic content.

```
[Fact]
public async Task TestCase()
{
    using (var store = GetDocumentStore())
    {
        using (var session = store.OpenAsyncSession())
        {
            await session.StoreAsync(new ConcatClass
            {
                Name = new Stub
                {
                    Name = "someName"
                }
            },"someId");
            await session.SaveChangesAsync();
            
            await session.StoreAsync(new ConcatClassWithDynamic
            {
                Name = new
                {
                    Name = "someName"
                }
            },"someId2");
            await session.SaveChangesAsync();
        }
        
        using (var session = store.OpenAsyncSession())
        {
           var doc = await session.LoadAsync<ConcatClass2>("someId");
           var name = doc.Name["Name"].ToString();
           Assert.Equal("someName",name);
        }    
        
        using (var session = store.OpenAsyncSession())
        {
            var doc = await session.LoadAsync<ConcatClass2>("someId2");
            var name = doc.Name["Name"].ToString();
            Assert.Equal("someName",name);
        }  
        
    }
}

private class ConcatClass
{
    public string Id { get; set; }
    public Stub Name { get; set; }
}

private class ConcatClass2
{
    public string Id { get; set; }
    public JObject Name { get; set; }
}    

private class ConcatClassWithDynamic
{
    public string Id { get; set; }
    public dynamic Name { get; set; }
}

private class Stub
{
    public string Name { get; set; }
}
```

**Comment by ayende** [comment, marked as answer]
Your approach doesn't really make sense, I'm afraid. You'll end up doing a LOT more work.
Also, a JSON string (as a string) means that you wouldn't be able to operate on the data on the server side.

Instead, use something like:


```
public class Foo
{
    public string Name;

    public dynamic DynamicValue1;
    public JObject DynamicValue2;
    public Dictionary<string,string> DynamicValue3;
    public Dictionary<string,object> DynamicValue4;
    public object DynamicValue5;
}
```

You can use any of the above options to store dynamic data, and they will allow you to process them on the server side and be efficient on the client.

**Reply by ben-jacobs** [reply]
Hi Oren (and Omer)

Yes I want it stored at DB level in the same way as a dynamic (for all the benefits that you mention) - but unfortunately I'm not starting with an object/dynamic. I am reading JSON-L files and want to store the JSON object on each line as a dynamic property of a new document (each line has a new document) - without the overhead of deserializing to an object. 

Unless I'm missing something, your suggestion is assuming I am starting with an object (ie the deserialized form). 

It wasn't overly important, I was just feeling out if anyone had done something similar. I'm dealing with streaming from largish compressed source files (multiple TB range) and it's not easy in this particular case to distribute work so I was trying to optimise as much as I could. But sometimes you just have to pick your battles.

Thanks for your help.

Ben

**Reply by ayende** [reply]
I'm not sure what the relevant costs are. I would suggest just doing the deserialization and seeing if the cost is acceptable.

Note that we provide a way for you to handle the serialization yourself, using:

https://github.com/ravendb/ravendb/blob/c512fb46d50385be2250d39881157738c5bb9361/src/Raven.Client/Json/Serialization/IJsonSerializer.cs#L7

You can plug this via:

https://github.com/ravendb/ravendb/blob/cd364100b1d172d0e2620c661ff31cbb93ffb648/src/Raven.Client/Json/Serialization/ISerializationConventions.cs#L14

That will allow you to have fine grained control of the output, in particular, take a look here:

https://github.com/ravendb/ravendb/blob/c512fb46d50385be2250d39881157738c5bb9361/src/Raven.Client/Json/Serialization/IJsonWriter.cs#L32

That can allow you to just send the string directly.

**Reply by ben-jacobs** [reply]
Thanks Oren - I know it's a bit of an usual path, but I'm going to persist for a few more seconds :). 

I'm seeing a 2-3x increase in throughput with this approach which equates to several days of time saved over a large dataset. In the convention I'm basically checking if the entity is BlittableJsonReaderObject and writing it using WriteJsonToAsync.
``
conventions.BulkInsert.TrySerializeEntityToJsonStream = (entity, metadata, writer) =>
			{
				requestExecutor ??= store.GetRequestExecutor(store.Database);
				requestExecutor.ContextPool.AllocateOperationContext(out JsonOperationContext context);

				if (entity is BlittableJsonReaderObject)
				{
					(entity as BlittableJsonReaderObject).WriteJsonToAsync(writer.BaseStream);
					return true;
				}
				else
				{
					using (var json = store.Conventions.Serialization.DefaultConverter.ToBlittable(entity, metadata, context, jsonSerializer ??= store.Conventions.Serialization.CreateSerializer()))
					{
						json.WriteJsonToAsync(writer.BaseStream);
						return true;
					}
				}
			};
``
This works well, uses low memory, and has a significant improvement in throughput - but obviously ignores any metadata passed in the StoreAsync function. In the source I can see where ToBlittable... takes a metadata dictionary in creating the blittable. But not much else that might assist with adding metadata to an already existing BlittableJsonObject.

I'm pretty sure I know the answer (there isn't one) but are you aware of any way to add metadata to a BlittableJsonObject once it has been created?

I know it's not standard, and I'll perhaps fall back to a faster serialization library than Json.NET if this raw BlittableJsonObject approach goes nowhere or investigate the serializer approach you outlined previously.

Ben

**Comment by haludi** [comment]
Hi,

What kind of metadata properties do you want to add?
If its RavenDB internal usage properties like @collection you should use
the convention -
https://ravendb.net/docs/article-page/5.1/csharp/client-api/configuration/identifier-generation/global#findcollectionname-and-findcollectionnamefordynamic

For costume metadata properties you can do the following

```
[Fact]
public void TestCase()
{
    IJsonSerializer jsonSerializer = null;
    RequestExecutor requestExecutor = null;
    using (var store = GetDocumentStore(new Options
    {
        ModifyDocumentStore = store =>
store.Conventions.BulkInsert.TrySerializeEntityToJsonStream = (entity,
metadata, writer) =>
        {

            requestExecutor ??= store.GetRequestExecutor(store.Database);
            requestExecutor.ContextPool.AllocateOperationContext(out
JsonOperationContext context);

            if (entity is BlittableJsonReaderObject bliEntity)
            {
                bliEntity.Modifications = new DynamicJsonValue();

bliEntity.Modifications[Constants.Documents.Metadata.Key] = new
DynamicJsonValue
                {
                    ["SomePropName"] = "SomePropValue"
                };

                // bliEntity.WriteJsonToAsync(writer.BaseStream);
                context.ReadObject(bliEntity,
"").WriteJsonToAsync(writer.BaseStream);
                return true;
            }
            else
            {
                using (var json =
store.Conventions.Serialization.DefaultConverter.ToBlittable(entity,
metadata, context, jsonSerializer ??=
store.Conventions.Serialization.CreateSerializer()))
                {
                    json.WriteJsonToAsync(writer.BaseStream);
                    return true;
                }
            }
        }
    }))
    {
        using (var bulkInsert = store.BulkInsert())
        using (store.GetRequestExecutor().ContextPool.AllocateOperationContext(out
var context))
        {
            var entity = new DynamicJsonValue();
            var bliEntity = context.ReadObject(entity, "");
            bulkInsert.Store(bliEntity);
        }
        WaitForUserToContinueTheTest(store);
    }
}
```

Best regards,
Igal

**Reply by ben-jacobs** [reply]
Thanks Igal - the custom metadata was exactly what I needed (at the moment). 

It would have taken me forever to stumble across the Modifications property and how to use it (not sure if this more low-level approach is really encouraged through any docs though :P).

I know this approach outlined above is a little out of the ordinary but I am truly amazed by the throughput difference in my bulk insert scenario (of course the difference is primarily cutting out all of the additional serialization work, but it's wonderful that these sort of options are in the product). 

Thanks again,
Ben

**Reply by ayende** [reply]
Once you are done, I would love to hear what numbers you were able to get
